crc32.c

```diff
@@ -217,22 +282,31 @@ unsigned long ZEXPORT crc32_z(crc, buf, len)
 
         endian = 1;
         if (*((unsigned char *)(&endian)))
-            return crc32_little(crc, buf, len);
+            return crc32_little;
         else
-            return crc32_big(crc, buf, len);
+            return crc32_big;
     }
 #endif /* BYFOUR */
-    crc = crc ^ 0xffffffffUL;
-    while (len >= 8) {
-        DO8;
-        len -= 8;
-    }
-    if (len) do {
-        DO1;
-    } while (--len);
-    return crc ^ 0xffffffffUL;
+
+    return crc32_table_lookup;
 }
 
+#if !defined(Z_IFUNC_ASM) && !defined(Z_IFUNC_NATIVE)
+
+unsigned long ZEXPORT crc32_z(crc, buf, len)
+    unsigned long crc;
+    const unsigned char FAR *buf;
+    z_size_t len;
+{
+    static unsigned long ZEXPORT (*crc32_func)(unsigned long, const unsigned char FAR *, z_size_t) = NULL;
+
+    if (!crc32_func)
+        crc32_func = crc32_z_ifunc();
```

Comment:
Another potential issue: would this work in a multithreaded application? Seems a bit race-condition-prone.

Maybe use a pthread_lock or pthread_once for the initialization?

Reply:
I would recommend to validate this with a highly parallel application (i.e. Chrome) or pigz (https://github.com/madler/pigz).

Reply:
Good question about thread safety. I put it to https://github.com/Dead2/zlib-ng/issues/147 who rely on function pointers extensively. Though they would have the equivalent of a crc32_func pointing at crc32_z_ifunc which modified the function pointer. I think I could do an approach that way which would avoid the high 32bits of crc32_func being modified, checked in another thread to be !0 and then called without the lower 32bits being set. Having said that I was unable to generate a normal ASM code that did it that way.

Does pthread_once work for initialising in static libraries? I'll try to avoid locks in the path as that seems more harmful than function pointers.

Reply:
@grooverdan they should work as zlib is built as a static lib in Chromium.